src/lib.rs

```diff
@@ -457,34 +520,224 @@ impl Cascade {
+            self.to_exclude.push(generator);
+        }
+    }
+
+    /// `exclude_threaded` is like `exclude` but it stores false positives in a caller-owned
+    /// `ExcludeSet`. This allows the caller to exclude items in parallel.
+    pub fn exclude_threaded(&self, exclude_set: &mut ExcludeSet, item: Vec<u8>) {
+        exclude_set.size += 1;
+        let mut generator = CascadeIndexGenerator::new(self.hash_algorithm, item);
+        if self.filters[0].has(&mut generator, &self.salt) {
+            exclude_set.set.push(generator);
+        }
+    }
+
+    /// `collect_exclude_set` merges an `ExcludeSet` into the internal storage of the CascadeBuilder.
+    pub fn collect_exclude_set(&mut self, exclude_set: &mut ExcludeSet) {
+        match self.status {
+            BuildStatus(0, ref mut cap) if *cap >= exclude_set.size => *cap -= exclude_set.size,
+            _ => panic!("capacity violation"),
+        }
+        self.to_exclude.append(&mut exclude_set.set);
+    }
+
+    fn push_layer(&mut self) -> Result<(), CascadeError> {
+        // At even layers we encode elements of to_include. At odd layers we encode elements of
+        // to_exclude. In both cases, we track false positives by filtering the complement of the
+        // encoded set through the newly produced bloom filter.
+        let at_even_layer = self.filters.len() % 2 == 0;
+        let (to_encode, to_filter) = match at_even_layer {
+            true => (&mut self.to_include, &mut self.to_exclude),
+            false => (&mut self.to_exclude, &mut self.to_include),
+        };
+
+        let mut bloom = Bloom::new_crlite_bloom(to_encode.len(), to_filter.len(), false);
+
+        // temporarily take self.salt since we need an immutable reference for bloom.insert
+        let salt = std::mem::take(&mut self.salt);
+
+        to_encode.iter_mut().for_each(|x| {
+            x.next_layer();
+            bloom.insert(x, &salt)
+        });
+
+        let mut delta = to_filter.len();
+        to_filter.retain_mut(|x| {
+            x.next_layer();
+            bloom.has(x, &salt)
+        });
+        delta -= to_filter.len();
+
+        if delta == 0 {
+            // Check for collisions between the |to_encode| and |to_filter| sets.
+            // The implementation of PartialEq for CascadeIndexGenerator will successfully
+            // identify cases where the user called |include(item)| and |exclude(item)| for the
+            // same item. It will not identify collisions in the underlying hash function.
+            for x in to_encode.iter_mut() {
+                if to_filter.contains(x) {
+                    return Err(CascadeError::Collision);
+                }
+            }
+        }
+
+        self.salt = salt;
+        self.filters.push(bloom);
+        Ok(())
+    }
+
+    pub fn finalize(mut self) -> Result<Box<Cascade>, CascadeError> {
+        match self.status {
+            BuildStatus(0, 0) => (),
+            _ => panic!("capacity violation"),
```

Comment:
I feel like since these functions return `Result`s anyway, we should be able to handle these situations as `Err`s, but it's not a big deal.